fix: enable running in incognito mode for recorder

aki-colt opened the pull request:
fix the  issue #70 . Enable recorder running in incognito mode.

ruifigueira commented on src/server/crx.ts:
```diff
@@ -138,21 +138,17 @@ export class Crx extends SdkObject {
-
-    const incognitoTabId = await incognitoTabIdPromise;
+      // do not create but find exists one
+      chrome.tabs.query({windowId: windowId}, (tabIds) => {
```
It's better to use the async version of `chrome.tabs.query` instead of the callback

ruifigueira commented on src/server/crx.ts:
```diff
@@ -138,21 +138,17 @@ export class Crx extends SdkObject {
+      // do not create but find exists one
+      chrome.tabs.query({windowId: windowId}, (tabIds) => {
+        if (!tabIds) {
+          chrome.tabs.create({ url: 'about:blank', windowId });
```
It returns a promise, we should `await` or alternatively add `.catch(() => {})`

ruifigueira commented on examples/recorder-crx/src/background.ts:
```diff
@@ -104,6 +104,13 @@ async function getCrxApp() {
     return;
+   // if the tab is incognito, chek if can be started in incognito mode.
+  if (tab.incognito) {
+    if (!await chrome.extension.isAllowedIncognitoAccess() || !settings.playInIncognito) {
```
We cannot have async calls at this point, check line 118 comment (basically sidepanel won't open if it's called after an async call, it must be called synchronously with a user gesture)

ruifigueira commented:
@aki-colt I pulled your branch, but it was not working.

I added these changes, it's now working: 3be0e367f381ec3c7647526125e963536afd8a71

However, there's still some issues: for instance, if I open recorder in incognito, then close incognito mode and try to open in the default browser context, it fails.

aki-colt commented:
Thanks for the review . I will merge your branch and try to figure out what is happening.

aki-colt commented:
I think I found the issue. In Incognito mode, we modified the browser's context. However, if we launch in default mode next time, we don't reinitialize ```crxApp```, meaning the browser's context remains in the modified state, which might cause the issue. I changed ```getCrxApp``` function to initialize ```crxApp``` on every launch, and it resolved the issue. 
```ts
async function getCrxApp(incognito: boolean) {
  // always start a new crx app.
  await settingsInitializing;

  crxAppPromise = crx.start({ incognito }).then(crxApp => {
    crxApp.recorder.addListener('hide', async () => {
      await crxApp.detachAll();
    });
    crxApp.recorder.addListener('modechanged', async ({ mode }) => {
      await Promise.all([...attachedTabIds].map(tabId => changeAction(tabId, mode)));
    });
    crxApp.addListener('attached', async ({ tabId }) => {
      attachedTabIds.add(tabId);
      await changeAction(tabId, crxApp.recorder.mode());
    });
    crxApp.addListener('detached', async tabId => {
      attachedTabIds.delete(tabId);
      await changeAction(tabId, 'detached');
    });
    setTestIdAttributeName(settings.testIdAttributeName);
    return crxApp;
  });
  return await crxAppPromise;
}
```
But this is obviously not an elegant solution. What do you think? @ruifigueira

ruifigueira commented:
I see your point, good call!

Nevertheless, I think it's enough to just close and unset `crxAppPromise` when recorder is hidden:

```js
      crxApp.recorder.addListener('hide', async () => {
        // will create new crxApp next time
        await crxApp.close();
        crxAppPromise = undefined;
      });
```

I did a very quick test on my side and seems to work as expected.

aki-colt commented:
Great, pushed a new commit for this.

ruifigueira commented on examples/recorder-crx/src/background.ts:
```diff
@@ -73,57 +78,65 @@ async function changeAction(tabId: number, mode?: CrxMode | 'detached') {
-      });
-      setTestIdAttributeName(settings.testIdAttributeName);
+async function getCrxApp(incognito: boolean) {
+  // always start a new crx app.
```
I would keep the previous code, and only start `crxApp` when `crxAppPromise` is undefined (hiding recorder will close it and set it as undefined)

ruifigueira commented on examples/recorder-crx/src/background.ts:
```diff
@@ -73,57 +78,65 @@ async function changeAction(tabId: number, mode?: CrxMode | 'detached') {
   const crxApp = await getCrxApp(tab.incognito);
 
   try {
+    // await crxApp.attach(tab.id!);
```
Remove this comment

ruifigueira commented:
I merged the changes but I still need to look into the tests, the ones related with incognito mode are now failing

aki-colt commented:
In incognito mode, it doesn't always fail. After some debugging, I found that sometimes the `_waitConnect()` method of class `SidepanelRecorderWindow` fails (located in `src/server/recorder/sidepanelRecorderWindow.ts`). The failure is due to the `onConnect` listener for the port not being triggered. This is likely because we attempted to connect before setting up the listener (connect is in the `useEffect` of `examples/recorder-crx/src/crxRecorder.tsx`, we first launch the sidepanel and then register the listener). I now added a 500ms delay in `crxRecorder` before establishing the connection, which makes it always work in default and incognito mode. However, I wonder why it always works in default mode🤔. This is the `useEffect` code I modified in `crxRecorder`.

```ts
  React.useEffect(() => {
    let port: any = undefined;
    setTimeout(() => {
      port = chrome.runtime.connect({ name: 'recorder' });
      const onMessage = (msg: any) => {
        if (!('type' in msg) || msg.type !== 'recorder')
          return;

        switch (msg.method) {
          case 'setPaused': setPaused(msg.paused); break;
          case 'setMode': setMode(msg.mode); break;
          case 'setSources': setSources(msg.sources); break;
          case 'resetCallLogs': setLog(new Map()); break;
          case 'updateCallLogs': setLog(log => {
            const newLog = new Map<string, CallLog>(log);
            for (const callLog of msg.callLogs) {
              callLog.reveal = !log.has(callLog.id);
              newLog.set(callLog.id, callLog);
            }
            return newLog;
          }); break;
          case 'setRunningFile': setRunningFileId(msg.file); break;
          case 'elementPicked': setElementPicked(msg.elementInfo, msg.userGesture); break;
        }
      };
      port.onMessage.addListener(onMessage);

      window.dispatch = async (data: any) => {
        port.postMessage({ type: 'recorderEvent', ...data });
        if (data.event === 'fileChanged')
          setSelectedFileId(data.params.file);
      };
      loadSettings().then(settings => {
        setSettings(settings);
        setSelectedFileId(settings.targetLanguage);
      }).catch(() => { });

      addSettingsChangedListener(setSettings);
    }, 500)

    return () => {
      removeSettingsChangedListener(setSettings);
      port.disconnect();
    };
  }, []);
```

ruifigueira commented:
Great job, @aki-colt ! I suspect it is because we set the default path for sidepanel in manifest.json:

```json
  "side_panel": {
    "default_path": "index.html"
  },
```

If we change it to some spash screen or `about:blank`, I think we'll no longer have that issue and that `setTimeout` will be no longer needed.

aki-colt commented:
Well, I'm not entirely sure if this approach is feasible, as [Chrome](https://developer.chrome.com/docs/extensions/reference/api/sidePanel) specifies that the Sidepanel's path must point to an existing HTML file.

ruifigueira commented:
We can just pass an empty html.

I'm fixing it in https://github.com/ruifigueira/playwright-crx/pull/75, in case you want to give it a try.
